File: First-Camp/leetcode/662. Maximum Width of Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def widthOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        queue = deque()
        ans = -200

        queue.append((root, 0))
        while queue:
            length = len(queue)
            right = queue[0][1]
            left = queue[-1][1]
            ans = max(ans, left - right + 1)
            i = 0
            while length:
                curr = queue.popleft()
                index = curr[1] - right
                if curr[0].left:
                    queue.append((curr[0].left, 2*index + 1) )
                if curr[0].right:
                    queue.append((curr[0].right, 2*index + 2))
                length -= 1
        return ans      

File: Second-Camp/leetcode/662. Maximum Width of Binary Tree.cpp
#include <bits/stdc++.h>
using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int widthOfBinaryTree(TreeNode* root) {
        if (!root) return 0;

        int maxWidth = 1;
        queue<pair<TreeNode*, int>> level;
        level.push({root, 0});

        while (!level.empty()) {
            int len = level.size();
            int left = level.front().second;
            int right = level.back().second;
            maxWidth = max(maxWidth, right - left + 1);
            while (len--) {
                pair<TreeNode*, int> curr = level.front();
                level.pop();

                int index = curr.second - right;
                if (curr.first->left) 
                    level.push({curr.first->left, (long long)2*index + 1 });
                if (curr.first->right) 
                    level.push({curr.first->right, (long long)2*index + 2 });
                
            }
        }
        return maxWidth;
    }
};